File: lib/storage.ts
import type { AppData } from "./types";
import { normalizeMission } from "./mission-utils";

const STORAGE_KEY = "daily_ticker_data";

export const EMPTY_APP_DATA: AppData = {
  profiles: [],
  activeProfileId: null,
  missions: [],
  dailyEntries: [],
  dailyMissions: [],
};

export function loadAppData(): AppData {
  if (typeof window === "undefined") return EMPTY_APP_DATA;
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return EMPTY_APP_DATA;
    const parsed = { ...EMPTY_APP_DATA, ...JSON.parse(raw) } as AppData;
    return {
      ...parsed,
      missions: parsed.missions.map(normalizeMission),
    };
  } catch {
    return EMPTY_APP_DATA;
  }
}

export function saveAppData(data: AppData): void {
  if (typeof window === "undefined") return;
  localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
}

File: mobile/lib/data/storage.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

import '../data/defaults.dart';
import '../models/types.dart';

const storageKey = 'daily_ticker_data';

const emptyAppData = AppData(
  profiles: [],
  activeProfileId: null,
  missions: [],
  dailyEntries: [],
  dailyMissions: [],
);

Future<AppData> loadAppData() async {
  try {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(storageKey);
    if (raw == null) return emptyAppData;

    final parsed = AppData.fromJson(jsonDecode(raw) as Map<String, dynamic>);

    return parsed.copyWith(
      missions: parsed.missions.map(normalizeMission).toList(),
    );
  } catch (_) {
    return emptyAppData;
  }
}

Future<void> saveAppData(AppData data) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.setString(storageKey, jsonEncode(data.toJson()));
}

String encodeAppData(AppData data) {
  return const JsonEncoder.withIndent('  ').convert(data.toJson());
}

AppData? parseAppDataJson(String raw) {
  try {
    final parsed = AppData.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    return parsed.copyWith(
      missions: parsed.missions.map(normalizeMission).toList(),
    );
  } catch (_) {
    return null;
  }
}
